cw/cw305/ceca.py

```diff
@@ -0,0 +1,574 @@
+        Number of remaining traces.
+    """
+    min_trace = mean_trace - max_std * std_trace
+    max_trace = mean_trace + max_std * std_trace
+    tasks = [
+        worker.filter_noisy_traces.remote(min_trace, max_trace)
+        for worker in workers
+    ]
+
+    running_cnt = 0
+    while tasks:
+        done, tasks = ray.wait(tasks)
+        running_cnt += ray.get(done[0])
+    return running_cnt
+
+
+def compute_mean_plaintext_traces(workers):
+    """Computes mean traces for all values of all plaintext bytes.
+
+    This function distributes work to ``TraceWorker`` instances and collects
+    their results.
+
+    Args:
+        workers: ``TraceWorker`` handles.
+
+    Returns:
+        A (16, 256, NUM_SAMPLES) array A, where A[i, j, :] is the mean of all
+        traces where plaintext byte i is j.
+    """
+    tasks = [
+        worker.count_and_sum_plaintext_traces.remote() for worker in workers
+    ]
+
+    running_cnt = None
+    running_sum = None
+    while tasks:
+        done, tasks = ray.wait(tasks)
+        cnt, sum_ = ray.get(done[0])
+        if running_cnt is None:
+            running_cnt = np.copy(cnt)
+            running_sum = np.copy(sum_)
+        else:
+            running_cnt += cnt
+            running_sum += sum_
+
+    return running_sum / running_cnt
+
+
+@timer()
+def compute_pairwise_diffs_and_scores(mean_traces):
+    """Computes pairwise differences and their confidence scores between key
+    bytes.
+
+    This function correlates mean plaintext traces to pick the most likely
+    differences.
+
+    Args:
+        mean_traces: A (16, 256, NUM_SAMPLES) array of mean plaintext traces.
+
+    Returns:
+        A (16, 16, 2) array A, where A[i, j, 0] is the difference with the
+        largest correlation coefficient between plaintext bytes i and j, and
+        A[i, j, 1] is the corresponding confidence score.
+    """
+    pairwise_diffs_scores = np.zeros((16, 16, 2))
+    # All possible values of plaintext byte a.
+    alphas = np.arange(256)
+    # All possible differences.
+    diffs = np.arange(256)[:, np.newaxis]
+    # Values of plaintext byte b for all possible differences: A (256, 256)
+    # array where betas[i, j] = diffs[i] ^ alphas[j]
+    betas = alphas ^ diffs
+    # Compute the most likely differences between all pairs of plaintext bytes.
+    for a in range(16):
+        for b in range(a + 1, 16):
+            corrcoefs = np.corrcoef(mean_traces[a], mean_traces[b])
+            # np.corrcoef returns the values that we want in the upper right
+            # quadrant.
+            diff_corrcoefs = corrcoefs[alphas, 256 + betas].sum(axis=1)
+            best_diff = diff_corrcoefs.argmax()
+            # TODO: Analyze the effect of /diff_corrcoefs.mean() below.
+            pairwise_diffs_scores[(a, b), (b, a)] = (best_diff,
+                                                     diff_corrcoefs[best_diff] /
+                                                     diff_corrcoefs.mean())
```

Comment:
Ah, here comes the normalization. I think you need this to make all the `pairwise_diffs_scores` comparable. There are pairs for which the correlation is almost flat but constantly higher than the highest peaks of another pair. You need the normalization to "downvote" this flat correlation for `find_best_diffs()`.

Reply:
I think so too but I left a TODO nevertheless because IIRC I was able to recover the key without this a couple of times and would like to check how the results look like when running batch attacks.